mlir/lib/Dialect/MIOpen/Transforms/AlignTiling.cpp

```diff
@@ -138,27 +142,83 @@ template <typename T> struct MILARewritePattern : public OpRewritePattern<T> {
     return nAlloc->getResult(0);
   }
 
+  Value makeBroadcast(PatternRewriter &b, MemRefType outType, Value inp,
+                      const AffineMapAttr &inpMap) const {
+    auto inpIdxMap = inpMap.getAffineMap();
+    if (!inpIdxMap.isIdentity()) {
+      auto loc = inp.getLoc();
+      auto inpType = inp.getType().template cast<MemRefType>();
+      auto inpShape = inpType.getShape();
+      auto outShape = outType.getShape();
+
+      uint32_t diff = outShape.size() - inpShape.size();
+      SmallVector<uint32_t> broadcastedDims;
+      if (diff) {
+        // expand dims (size = 1) in front
+        SmallVector<uint32_t, 8> expDims(diff);
+        std::iota(expDims.begin(), expDims.end(), 0);
+        SmallVector<int64_t, 8> expSizes(diff, 1);
+
+        miopen::BottomUpCTBuilder transform(b, inpShape, loc);
+        transform.expand(expDims, expSizes);
+
+        inp = b.create<miopen::TransformOp>(loc, inp, transform.get(),
+                                            inpType.getMemorySpaceAsInt());
+
+        inpType = inp.getType().template cast<MemRefType>();
+        inpShape = inpType.getShape();
+        broadcastedDims = {expDims.begin(), expDims.end()};
+      } else {
+        inpIdxMap.isMinorIdentityWithBroadcasting(&broadcastedDims);
+      }
+
+      // 0. insert a broadcast miopen.transform
+      miopen::BottomUpCTBuilder transform(b, inpShape, loc);
+      transform.broadcast(broadcastedDims, outType.getShape());
```

Comment:
Yeah, I don't like this. The broadcast should specify exactly which dimensions it applies to and the rest should be passed through. Not doing this makes the code hard to reason about.

Reply:
The broadcast call here passes the dims that would be broadcast but it also needs the upper 'result' dim sizes since the builder takes the lower sizes only.

Reply:
But you're not broadcasting every dimension, while passing every size